sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/FilterEstimation.scala

```diff
@@ -0,0 +1,555 @@
+      logDebug("[CBO] No statistics for " + attrRef)
+      return 1.0
+    }
+    val aColStat = mutableColStats(attrRef.exprId)
+    val rowCountValue = plan.child.statistics.rowCount.get
+    val nullPercent: BigDecimal =
+      if (rowCountValue == 0) 0.0
+      else BigDecimal(aColStat.nullCount)/BigDecimal(rowCountValue)
+
+    if (update) {
+      val newStats =
+        if (isNull) aColStat.copy(distinctCount = 0, min = None, max = None)
+        else aColStat.copy(nullCount = 0)
+
+      mutableColStats += (attrRef.exprId -> newStats)
+    }
+
+    val percent =
+      if (isNull) nullPercent.toDouble
+      else {
+        /** ISNOTNULL(column) */
+        1.0 - nullPercent.toDouble
+      }
+
+    percent
+  }
+
+  /**
+   * Returns a percentage of rows meeting a binary comparison expression.
+   *
+   * @param op a binary comparison operator uch as =, <, <=, >, >=
+   * @param attrRef an AttributeReference (or a column)
+   * @param literal a literal value (or constant)
+   * @param update a boolean flag to specify if we need to update ColumnStat of a given column
+   *               for subsequent conditions
+   * @return a doube value to show the percentage of rows meeting a given condition
+   */
+  def evaluateBinary(
+      op: BinaryComparison,
+      attrRef: AttributeReference,
+      literal: Literal,
+      update: Boolean)
+    : Double = {
+    if (!mutableColStats.contains(attrRef.exprId)) {
+      logDebug("[CBO] No statistics for " + attrRef)
+      return 1.0
+    }
+
+    // Make sure that the Date/Timestamp literal is a valid one
+    attrRef.dataType match {
+      case DateType =>
+        val dateLiteral = DateTimeUtils.stringToDate(literal.value.asInstanceOf[UTF8String])
+        if (dateLiteral.isEmpty) {
+          logDebug("[CBO] Date literal is wrong, No statistics for " + attrRef)
+          return 1.0
+        }
+      case TimestampType =>
+        val tsLiteral = DateTimeUtils.stringToTimestamp(literal.value.asInstanceOf[UTF8String])
+        if (tsLiteral.isEmpty) {
+          logDebug("[CBO] Timestamp literal is wrong, No statistics for " + attrRef)
+          return 1.0
+        }
+      case _ =>
+    }
+
+    op match {
+      case EqualTo(l, r) => evaluateEqualTo(attrRef, literal, update)
+      case _ =>
+        attrRef.dataType match {
+          case _: NumericType | DateType | TimestampType =>
+            evaluateBinaryForNumeric(op, attrRef, literal, update)
+          case StringType | BinaryType =>
+
+            // TODO: It is difficult to support other binary comparisons for String/Binary
+            // type without min/max and advanced statistics like histogram.
+
+            logDebug("[CBO] No statistics for String/Binary type " + attrRef)
+            return 1.0
+        }
+    }
+  }
+
+  /**
+   * This method converts a numeric or Literal value of numeric type to a BigDecimal value.
+   * If isNumeric is true, then it is a numeric value.  Otherwise, it is a Literal value.

```

Comment:
For most predicates, parser will make a constant value as a Literal.  However, for a predicate such as "key IN (3, 4, 5)", I found that values 3, 4, 5 are not Literal objects.  They are just numeric values.